sql/expression/in.go

```diff
@@ -142,3 +147,84 @@ func (in *InTuple) Children() []sql.Expression {
 func NewNotInTuple(left sql.Expression, right sql.Expression) sql.Expression {
 	return NewNot(NewInTuple(left, right))
 }
+
+// HashInTuple is an expression that checks an expression is inside a list of expressions using a hashmap.
+type HashInTuple struct {
+	InTuple
+	cmp       map[interface{}]sql.Expression
+	hasNull   bool
+	rightType sql.Type
+}
+
+// NewHashInTuple creates an InTuple expression.
+func NewHashInTuple(left, right sql.Expression) (*HashInTuple, error) {
+	cmp, hasNull, t, err := newInMap(right)
+	if err != nil {
+		return nil, err
+	}
+
+	return &HashInTuple{InTuple: *NewInTuple(left, right), cmp: cmp, hasNull: hasNull, rightType: t}, nil
+}
+
+// Eval implements the Expression interface.
+func (hit *HashInTuple) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	if hit.hasNull {
+		return nil, nil
+	}
+
+	leftElems := sql.NumColumns(hit.Left().Type().Promote())
+	left, err := hit.Left().Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if left == nil {
+		return nil, nil
+	}
+
+	left, err = hit.rightType.Convert(left)
+	if err != nil {
+		return nil, err
+	}
+
+	right, ok := hit.cmp[left]
+	if !ok {
+		return false, nil
+	}
+	if sql.NumColumns(right.Type()) != leftElems {
+		return nil, sql.ErrInvalidOperandColumns.New(leftElems, sql.NumColumns(right.Type()))
+	}
+
+	return true, nil
+}
+
+func (hit *HashInTuple) String() string {
+	return fmt.Sprintf("(%s HASH IN %s)", hit.Left(), hit.Right())
+}
+
+func (hit *HashInTuple) DebugString() string {
+	return fmt.Sprintf("(%s HASH IN %s)", sql.DebugString(hit.Left()), sql.DebugString(hit.Right()))
+}
+
+func newInMap(expr sql.Expression) (map[interface{}]sql.Expression, bool, sql.Type, error) {
+	elements := make(map[interface{}]sql.Expression)
```

Comment:
Probably should manually hash everything in here and make this `map[uint64]sql.Expression`